src/main/webapp/app/course/manage/course-management-exercises.component.html

```diff
@@ -4,25 +4,25 @@
             <h3>
                 <span>
                     {{ course.title }}
-                    - {{ getExerciseCount() }}
+                    - {{ exerciseCount }}
                     {{ 'artemisApp.course.exercises' | artemisTranslate }}
                 </span>
             </h3>
         </div>
         <div class="col text-end align-self-center">
-            <button id="toggleSearchButton" class="btn btn-primary" type="button" (click)="toggleSearch()">
+            <button id="toggleSearchButton" class="btn btn-primary" type="button" (click)="showSearch = !showSearch">
                 {{ 'artemisApp.course.search.toggleButton.' + showSearch | artemisTranslate }}
             </button>
         </div>
     </div>
 
     <jhi-course-management-exercises-search *ngIf="showSearch" (exerciseFilter)="exerciseFilter = $event"></jhi-course-management-exercises-search>
```

Comment:
I'm not really sure if your complicated changes are necessary.
The only way your numbers can change is through filtering, right?
If so, you could just call a method here with the filter which sets all parameters accordingly. Then you don't need all the getters and setters. You can set the attributes readonly then which will resolve the issue of accidental updates from outside the class.

Reply:
Updating the attributes when the filter changes is actually not easily possible since the exerciseCounts and filteredExerciseCounts are calculated in the exercise.component.ts files. These are then outputted individually to the parent component. So the only way I see to use a single method that is executed when the filter changes would be to move the exercise loading and filtering from the exercise components to the course-management-exercises component which would be even worse.
And I also can't just use Angulars OnChanges interface as this would result in the same problem that this refactoring was supposed to prevent in the first place (Updating everything when nothing really changes).

Reply:
This is quite an old thread but could you maybe check if this works? I think this would resolve the issue:
https://stackoverflow.com/questions/39331125/angular2-call-function-when-the-input-changes
=>  you keep `[exerciseFilter]="exerciseFilter" for your exercises but in the TS file you have a setter for `exerciseFilter` instead of an attribute annotated with `@Input`. This way you should be able to filter on the change of only this specific input.

Reply:
The thing is that this would also not work.
The child components change the parent component, as they all emit their exerciseCounts.
And this change happens **after** the filter in the parent component has been changed.
So if I would do an update when the filter in the parent changes the changes from the childs would happen after this and would therefore not be applied.
I need to listen to the update from every child to check if I should display the no result message.

Reply:
But couldn't you then just add similar setters for `filteredXYZExerciseCount` in the course-management-exercises.component.ts`?
This should resolve this issue and would, in my opinion, still be better than what we currently have. We probably don't even need getters for the attributes. I don't think anyone needs them.

Or maybe the better solution:
Use `@ViewChild` to get the reference to the child components in our Typescript code. You can then tell them directly to filter, directly retrieve the new amount and recalculate whether to display the warning or not.
I think this would make even more sense as the action "filter" is not an internal action (e.g. compared to the complaints where the user submits the child component form and, therefore, the event has to be propagated upwards to the parent in order to update the layout.

Reply:
> But couldn't you then just add similar setters for filteredXYZExerciseCount in the course-management-exercises.component.ts`?
This should resolve this issue and would, in my opinion, still be better than what we currently have. We probably don't even need getters for the attributes. I don't think anyone needs them.

I have already added setters for filteredXYZExerciseCounts.
And I did only add a getter for the ExerciseFilter, as this one is necessary.
So I am not sure what you want me to do now.

The @ViewChild method could work, but it would result in more complex code. Also, at least to me, it looks like it would not fit into Angular's usual way of passing attributes with @Input and @Output.

Reply:
Regarding the first one, I was referring to setters annotated with `@Input`

Regarding the second one, we already use `@ViewChild` throughout our system so I don't think this is an issue
I also think that these are not exclusive. `@Input` and `@Output` relate to changes that do not depend on other changes/components. Also a way to do quick changes in the template. `@ViewChild` is basically a more powerful way because it allows doing changes that depend on other changes/components (as we need right here) but is limited to the Typescript code. Which is usually not an issue since the code is more complicated anyway.

Reply:
Refactoring is out of scope. For the trade-off readability vs coding guidelines & performance, the readability wins as it hadn't too much impact until now.

Future refactoring might be nice but for now, we stick with the function calls to not make things too complicated for this PR.